util/process_wrapper/flags.rs

```diff
@@ -0,0 +1,270 @@
+        if self.repeated.contains_key(name) {
+            panic!("argument \"{}\" already defined as repeated flag", name)
+        }
+        self.single.insert(
+            name.to_owned(),
+            FlagDef::<'a, String> {
+                name: name.to_owned(),
+                help: help.to_owned(),
+                data,
+            },
+        );
+    }
+
+    pub(crate) fn define_repeated_flag(
+        &mut self,
+        name: &str,
+        help: &str,
+        data: &'a mut Option<Vec<String>>,
+    ) {
+        if self.single.contains_key(name) {
+            panic!("argument \"{}\" already defined as flag", name)
+        }
+        self.repeated.insert(
+            name.to_owned(),
+            FlagDef::<'a, Vec<String>> {
+                name: name.to_owned(),
+                help: help.to_owned(),
+                data,
+            },
+        );
+    }
+
+    fn help(&self, program_name: String) -> String {
+        let single = self.single.values().map(|fd| fd.to_string());
+        let repeated = self.repeated.values().map(|fd| fd.to_string());
+        let mut all: Vec<String> = single.chain(repeated).collect();
+        all.sort();
+
+        let mut help_text = String::new();
+        writeln!(
+            &mut help_text,
+            "Help for {}: [options] -- [extra arguments]",
+            program_name
+        )
+        .unwrap();
+        for line in all {
+            writeln!(&mut help_text, "\t{}", line).unwrap();
+        }
+        help_text
+    }
+
+    pub(crate) fn parse(mut self, mut argv: Vec<String>) -> Result<ParseOutcome, FlagParseError> {
+        let program_name = take(argv.first_mut().unwrap());
+        // To check if a non-repeated flag has been set already.
+        let mut set = HashSet::<String>::new();
+
+        for (flag, mut args) in iter_flags(argv.into_iter().skip(1))? {
+            if flag == "--help" {
+                return Ok(ParseOutcome::Help(self.help(program_name)));
+            }
+            if flag == "--" {
+                return Ok(ParseOutcome::Parsed(args));
+            }
+            if args.is_empty() {
+                return Err(FlagParseError::ValueMissing(flag.clone()));
+            }
+            if let Some(e) = self.single.get_mut(&flag) {
+                if args.len() > 1 || set.contains(&flag) {
+                    return Err(FlagParseError::ProvidedMultipleTimes(flag.clone()));
+                }
+                let arg = args.first_mut().unwrap();
+                set.insert(flag);
+                *e.data = Some(take(arg));
+                continue;
+            }
+            if let Some(e) = self.repeated.get_mut(&flag) {
+                e.data.get_or_insert_with(Vec::new).append(&mut args);
+                continue;
+            }
+            return Err(FlagParseError::UnkownFlag(flag));
+        }
+        Ok(ParseOutcome::Parsed(vec![]))
+    }
+}
+
+// Returns an iterator over Vecs [--foo, arg1, arg2, ... argN].
+fn iter_flags(
```

Comment:
Github just decided to throw away my comment because the diff was not valid anymore, and let me send my other comments. Yay I get to write this from scratch again. -.-

So, I find it a bit hard to follow the code and I was wondering if code like this would be more readable:

```
let mut argv_iter = argv.into_iter().peekable();

while let Some(_) = argv_iter.peek() {
	let (flag, values, rest) = consume_next_flag(argv_iter);
	...
	argv_iter = rest;
}
```

WDYT?

Reply:
I simplified it a bit further. WDYT?